README.md
# ExAws.ECS

Service module for https://github.com/ex-aws/ex_aws

## Installation

The package can be installed by adding `ex_aws_ecs` to your list of dependencies in `mix.exs`along with `:ex_aws` and your preferred JSON codec / http client.

```elixir
def deps do
  [
    {:ex_aws, "~> 2.0"}
    {:ex_aws_ecs, "~> 0.1.0"}
    {:poison, "~> 3.0"},
    {:hackney, "~> 1.9"}
  ]
end
```

Documentation can be found at https://hexdocs.pm/ex_aws_ecs.

## License

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# v 0.1.0

Initial Release

# v 0.1.1

Bugfix documentation

test/ex_aws/ex_aws_ecs_test.exs
defmodule ExAws.ECSTest do
  use ExUnit.Case, async: true
  doctest ExAws.ECS

  test "create_cluster" do
    req = ExAws.ECS.create_cluster("new_cluster")

    assert req.data == %{
             "Action" => "CreateCluster",
             "Version" => "2014-11-13",
             "clusterName" => "new_cluster"
           }

    assert req.headers == expected_headers("CreateCluster")
  end

  test "create_service" do
    req = ExAws.ECS.create_service("new_service", "arn:::task:def", 2)

    assert req.data == %{
             "Action" => "CreateService",
             "Version" => "2014-11-13",
             "serviceName" => "new_service",
             "taskDefinition" => "arn:::task:def",
             "desiredCount" => 2
           }

    assert req.headers == expected_headers("CreateService")
  end

  test "delete_attributes" do
    req = ExAws.ECS.delete_attributes("name=string,value=string")

    assert req.data == %{
             "Action" => "DeleteAttributes",
             "Version" => "2014-11-13",
             "attributes" => "name=string,value=string"
           }

    assert req.headers == expected_headers("DeleteAttributes")
  end

  test "delete_cluster" do
    req = ExAws.ECS.delete_cluster("arn::cluster:id")

    assert req.data == %{
             "Action" => "DeleteCluster",
             "Version" => "2014-11-13",
             "cluster" => "arn::cluster:id"
           }

    assert req.headers == expected_headers("DeleteCluster")
  end

  test "delete_service" do
    req = ExAws.ECS.delete_service("arn::service:id")

    assert req.data == %{
             "Action" => "DeleteService",
             "Version" => "2014-11-13",
             "service" => "arn::service:id"
           }

    assert req.headers == expected_headers("DeleteService")
  end

  test "deregister_container_instance" do
    req = ExAws.ECS.deregister_container_instance("arn::container:id")

    assert req.data == %{
             "Action" => "DeregisterContainerInstance",
             "Version" => "2014-11-13",
             "containerInstance" => "arn::container:id"
           }

    assert req.headers == expected_headers("DeregisterContainerInstance")
  end

  test "deregister_task_definition" do
    req = ExAws.ECS.deregister_task_definition("arn::task:def")

    assert req.data == %{
             "Action" => "DeregisterTaskDefinition",
             "Version" => "2014-11-13",
             "taskDefinition" => "arn::task:def"
           }

    assert req.headers == expected_headers("DeregisterTaskDefinition")
  end

  test "describe_clusters" do
    req = ExAws.ECS.describe_clusters()

    assert req.data == %{
             "Action" => "DescribeClusters",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("DescribeClusters")
  end

  test "describe_container_instances" do
    req = ExAws.ECS.describe_container_instances(["id1", "id2"])

    assert req.data == %{
             "Action" => "DescribeContainerInstances",
             "Version" => "2014-11-13",
             "containerInstances" => ["id1", "id2"]
           }

    assert req.headers == expected_headers("DescribeContainerInstances")
  end

  test "describe_services" do
    req = ExAws.ECS.describe_services(["id1", "id2"])

    assert req.data == %{
             "Action" => "DescribeServices",
             "Version" => "2014-11-13",
             "services" => ["id1", "id2"]
           }

    assert req.headers == expected_headers("DescribeServices")
  end

  test "describe_tasks" do
    req = ExAws.ECS.describe_tasks("cluster", ["id1", "id2"])

    assert req.data == %{
             "Action" => "DescribeTasks",
             "Version" => "2014-11-13",
             "cluster" => "cluster",
             "tasks" => ["id1", "id2"]
           }

    assert req.headers == expected_headers("DescribeTasks")
  end

  test "discover_poll_endpoint" do
    req = ExAws.ECS.discover_poll_endpoint()

    assert req.data == %{
             "Action" => "DiscoverPollEndpoint",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("DiscoverPollEndpoint")
  end

  test "list_attributes" do
    req = ExAws.ECS.list_attributes("target_type_id")

    assert req.data == %{
             "Action" => "ListAttributes",
             "Version" => "2014-11-13",
             "targetType" => "target_type_id"
           }

    assert req.headers == expected_headers("ListAttributes")
  end

  test "list_clusters" do
    req = ExAws.ECS.list_clusters()

    assert req.data == %{
             "Action" => "ListClusters",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("ListClusters")
  end

  test "list_container_instances" do
    req = ExAws.ECS.list_container_instances()

    assert req.data == %{
             "Action" => "ListContainerInstances",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("ListContainerInstances")
  end

  test "list_services" do
    req = ExAws.ECS.list_services()

    assert req.data == %{
             "Action" => "ListServices",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("ListServices")
  end

  test "list_task_definition_families" do
    req = ExAws.ECS.list_task_definition_families()

    assert req.data == %{
             "Action" => "ListTaskDefinitionFamilies",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("ListTaskDefinitionFamilies")
  end

  test "list_task_definitions" do
    req = ExAws.ECS.list_task_definitions()

    assert req.data == %{
             "Action" => "ListTaskDefinitions",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("ListTaskDefinitions")
  end

  test "list_tasks" do
    req = ExAws.ECS.list_tasks("cluster")

    assert req.data == %{
             "Action" => "ListTasks",
             "Version" => "2014-11-13",
             "cluster" => "cluster"
           }

    assert req.headers == expected_headers("ListTasks")
  end

  test "put_attributes" do
    req = ExAws.ECS.put_attributes("name=string,value=string")

    assert req.data == %{
             "Action" => "PutAttributes",
             "Version" => "2014-11-13",
             "attributes" => "name=string,value=string"
           }

    assert req.headers == expected_headers("PutAttributes")
  end

  test "register_container_instances" do
    req = ExAws.ECS.register_container_instance()

    assert req.data == %{
             "Action" => "RegisterContainerInstance",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("RegisterContainerInstance")
  end

  test "register_task_definition" do
    req = ExAws.ECS.register_task_definition("family", ["def1", "def2"])

    assert req.data == %{
             "Action" => "RegisterTaskDefinition",
             "Version" => "2014-11-13",
             "family" => "family",
             "containerDefinitions" => ["def1", "def2"]
           }

    assert req.headers == expected_headers("RegisterTaskDefinition")
  end

  test "run_task" do
    req = ExAws.ECS.run_task("task_def")

    assert req.data == %{
             "Action" => "RunTask",
             "Version" => "2014-11-13",
             "taskDefinition" => "task_def"
           }

    assert req.headers == expected_headers("RunTask")
  end

  test "start_task" do
    req = ExAws.ECS.start_task("task_def", ["cinst1", "cinst2"])

    assert req.data == %{
             "Action" => "StartTask",
             "Version" => "2014-11-13",
             "taskDefinition" => "task_def",
             "containerInstances" => ["cinst1", "cinst2"]
           }

    assert req.headers == expected_headers("StartTask")
  end

  test "stop_task" do
    req = ExAws.ECS.stop_task("task_id")

    assert req.data == %{
             "Action" => "StopTask",
             "Version" => "2014-11-13",
             "task" => "task_id"
           }

    assert req.headers == expected_headers("StopTask")
  end

  test "submit_container_state_change" do
    req = ExAws.ECS.submit_container_state_change()

    assert req.data == %{
             "Action" => "SubmitContainerStateChange",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("SubmitContainerStateChange")
  end

  test "submit_task_state_change" do
    req = ExAws.ECS.submit_task_state_change()

    assert req.data == %{
             "Action" => "SubmitTaskStateChange",
             "Version" => "2014-11-13"
           }

    assert req.headers == expected_headers("SubmitTaskStateChange")
  end

  test "update_container_agent" do
    req = ExAws.ECS.update_container_agent("cinst1")

    assert req.data == %{
             "Action" => "UpdateContainerAgent",
             "Version" => "2014-11-13",
             "containerInstance" => "cinst1"
           }

    assert req.headers == expected_headers("UpdateContainerAgent")
  end

  test "update_container_instances_state" do
    req = ExAws.ECS.update_container_instances_state(["cinst1", "cinst2"], "STOPPED")

    assert req.data == %{
             "Action" => "UpdateContainerInstancesState",
             "Version" => "2014-11-13",
             "containerInstances" => ["cinst1", "cinst2"],
             "status" => "STOPPED"
           }

    assert req.headers == expected_headers("UpdateContainerInstancesState")
  end

  test "update_service" do
    req = ExAws.ECS.update_service("service_id", desired_count: 4)

    assert req.data == %{
             "Action" => "UpdateService",
             "Version" => "2014-11-13",
             "service" => "service_id",
             "desiredCount" => 4
           }

    assert req.headers == expected_headers("UpdateService")
  end

  defp expected_headers(action) do
    [
      {"x-amz-target", "AmazonEC2ContainerServiceV20141113.#{action}"},
      {"content-type", "application/x-amz-json-1.1"}
    ]
  end
end


mix.exs
defmodule ExAws.ECS.Mixfile do
  use Mix.Project

  @version "0.1.1"
  @url "https://github.com/davepersing/ex_aws_ecs"
  @maintainers [
    "Dave Persing"
  ]

  def project do
    [
      name: "ExAws.ECS",
      app: :ex_aws_ecs,
      version: @version,
      elixir: "~> 1.5",
      elixirc_paths: elixirc_paths(Mix.env()),
      package: package(),
      source_url: @url,
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      maintainers: @maintainers,
      description: "ECS Plugin for ExAws framework",
      homepage_url: @url,
      docs: docs(),
      deps: deps()
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger, :ex_aws]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc, "~> 0.16", only: :dev, runtime: false},
      {:credo, "~> 0.8.6", only: [:dev, :test]},
      {:hackney, ">= 0.0.0", only: [:dev, :test]},
      {:poison, ">= 0.0.0", only: [:dev, :test]},
      ex_aws()
    ]
  end

  def docs do
    [
      extras: ["README.md", "CHANGELOG.md"],
      source_ref: "v#{@version}"
    ]
  end

  defp ex_aws() do
    case System.get_env("AWS") do
      "LOCAL" -> {:ex_aws, path: "../ex_aws"}
      _ -> {:ex_aws, "~> 2.0.0"}
    end
  end

  defp package do
    [
      maintainers: @maintainers,
      licenses: ["MIT"],
      links: %{github: @url},
      files: ~w(lib) ++ ~w(CHANGELOG.md LICENSE mix.exs README.md)
    ]
  end
end


lib/ex_aws/ex_aws_ecs.ex
defmodule ExAws.ECS do
  @moduledoc """
  AWS ECS Service for ExAws
  """

  alias ExAws.Operation.JSON

  import ExAws.Utils, only: [camelize_keys: 1]

  @version "2014-11-13"
  @namespace "AmazonEC2ContainerServiceV20141113"

  def create_cluster(cluster_name) do
    request(%{
      "Action" => "CreateCluster",
      "clusterName" => cluster_name
    })
  end

  def create_service(service_name, task_definition, desired_count, opts \\ []) do
    %{
      "Action" => "CreateService",
      "serviceName" => service_name,
      "taskDefinition" => task_definition,
      "desiredCount" => desired_count
    }
    |> merge_opts(opts)
    |> request()
  end

  def delete_attributes(attributes, opts \\ []) do
    %{
      "Action" => "DeleteAttributes",
      "attributes" => attributes
    }
    |> merge_opts(opts)
    |> request()
  end

  def delete_cluster(cluster) do
    request(%{
      "Action" => "DeleteCluster",
      "cluster" => cluster
    })
  end

  def delete_service(service) do
    request(%{
      "Action" => "DeleteService",
      "service" => service
    })
  end

  def deregister_container_instance(container_instance, opts \\ []) do
    %{
      "Action" => "DeregisterContainerInstance",
      "containerInstance" => container_instance
    }
    |> merge_opts(opts)
    |> request()
  end

  def deregister_task_definition(task_definition) do
    request(%{
      "Action" => "DeregisterTaskDefinition",
      "taskDefinition" => task_definition
    })
  end

  def describe_clusters do
    request(%{"Action" => "DescribeClusters"})
  end

  def describe_container_instances(container_instances) do
    request(%{
      "Action" => "DescribeContainerInstances",
      "containerInstances" => container_instances
    })
  end

  def describe_services(services, opts \\ []) do
    %{
      "Action" => "DescribeServices",
      "services" => services
    }
    |> merge_opts(opts)
    |> request()
  end

  def describe_task_definition(task_definition) do
    request(%{
      "Action" => "DescribeTaskDefinition",
      "taskDefinition" => task_definition
    })
  end

  def describe_tasks(cluster, tasks, opts \\ []) do
    %{
      "Action" => "DescribeTasks",
      "cluster" => cluster,
      "tasks" => tasks
    }
    |> merge_opts(opts)
    |> request()
  end

  def discover_poll_endpoint(opts \\ []) do
    %{
      "Action" => "DiscoverPollEndpoint"
    }
    |> merge_opts(opts)
    |> request()
  end

  def list_attributes(target_type, opts \\ []) do
    %{
      "Action" => "ListAttributes",
      "targetType" => target_type
    }
    |> merge_opts(opts)
    |> request()
  end

  def list_clusters(opts \\ []) do
    %{
      "Action" => "ListClusters"
    }
    |> merge_opts(opts)
    |> request()
  end

  def list_container_instances(opts \\ []) do
    %{
      "Action" => "ListContainerInstances"
    }
    |> merge_opts(opts)
    |> request()
  end

  def list_services(opts \\ []) do
    %{
      "Action" => "ListServices"
    }
    |> merge_opts(opts)
    |> request()
  end

  def list_task_definition_families(opts \\ []) do
    %{
      "Action" => "ListTaskDefinitionFamilies"
    }
    |> merge_opts(opts)
    |> request()
  end

  def list_task_definitions(opts \\ []) do
    %{
      "Action" => "ListTaskDefinitions"
    }
    |> merge_opts(opts)
    |> request()
  end

  def list_tasks(cluster, opts \\ []) do
    %{
      "Action" => "ListTasks",
      "Version" => @version,
      "cluster" => cluster
    }
    |> merge_opts(opts)
    |> request()
  end

  def put_attributes(attributes, opts \\ []) do
    %{
      "Action" => "PutAttributes",
      "attributes" => attributes
    }
    |> merge_opts(opts)
    |> request()
  end

  def register_container_instance(opts \\ []) do
    %{
      "Action" => "RegisterContainerInstance"
    }
    |> merge_opts(opts)
    |> request()
  end

  def register_task_definition(family, container_definitions, opts \\ []) do
    %{
      "Action" => "RegisterTaskDefinition",
      "family" => family,
      "containerDefinitions" => container_definitions
    }
    |> merge_opts(opts)
    |> request()
  end

  def run_task(task_definition, opts \\ []) do
    %{
      "Action" => "RunTask",
      "taskDefinition" => task_definition
    }
    |> merge_opts(opts)
    |> request()
  end

  def start_task(task_definition, container_instances, opts \\ []) do
    %{
      "Action" => "StartTask",
      "taskDefinition" => task_definition,
      "containerInstances" => container_instances
    }
    |> merge_opts(opts)
    |> request()
  end

  def stop_task(task, opts \\ []) do
    %{
      "Action" => "StopTask",
      "task" => task
    }
    |> merge_opts(opts)
    |> request()
  end

  def submit_container_state_change(opts \\ []) do
    %{
      "Action" => "SubmitContainerStateChange"
    }
    |> merge_opts(opts)
    |> request()
  end

  def submit_task_state_change(opts \\ []) do
    %{
      "Action" => "SubmitTaskStateChange"
    }
    |> merge_opts(opts)
    |> request()
  end

  def update_container_agent(container_instance, opts \\ []) do
    %{
      "Action" => "UpdateContainerAgent",
      "containerInstance" => container_instance
    }
    |> merge_opts(opts)
    |> request()
  end

  def update_container_instances_state(container_instances, status, opts \\ []) do
    %{
      "Action" => "UpdateContainerInstancesState",
      "containerInstances" => container_instances,
      "status" => status
    }
    |> merge_opts(opts)
    |> request()
  end

  def update_service(service, opts \\ []) do
    %{
      "Action" => "UpdateService",
      "service" => service
    }
    |> merge_opts(opts)
    |> request()
  end

  defp merge_opts(params, opts) do
    opts
    |> normalize_opts()
    |> Map.merge(params)
  end

  defp request(%{"Action" => action} = params, opts \\ %{}) do
    params = Map.merge(params, %{"Version" => @version})

    JSON.new(
      :ecs,
      %{
        data: params,
        headers: [
          {"x-amz-target", "#{@namespace}.#{action}"},
          {"content-type", "application/x-amz-json-1.1"}
        ]
      }
      |> Map.merge(opts)
    )
  end

  # This JSON API requires lower camelCased parameters.
  defp normalize_opts(opts) do
    opts
    |> Enum.into(%{})
    |> camelize_keys()
    |> Enum.reduce(
      %{},
      fn {k, v}, acc ->
        Map.put(acc, to_lower_camel_case(k), v)
      end
    )
  end

  defp to_lower_camel_case(string) do
    {first, rest} = String.split_at(string, 1)

    String.downcase(first) <> rest
  end
end
